### DISCUSSION TITLE
coredns returning in cluster service ip's when queried from outside

### DISCUSSION THREAD

**Discussion by micksey9** [coredns/coredns#4898] [Q&A] [answered]
Hey all, I have k8s_external plugin configured for an external zone **example.org** and have also exposed coredns as load balancer service so that coredns can act as NS for this external zone. The issue is that if I query coredns from outside the cluster for a FQDN for an in cluster (cluster local) service then coredns is responding to that query with the in cluster service ip. 

Following is my corefile:
```
  Corefile: |
    .:53 {
        errors
        health {
            lameduck 5s
        }
        ready
        kubernetes cluster.local in-addr.arpa ip6.arpa {
          pods insecure
          fallthrough in-addr.arpa ip6.arpa
        }
        prometheus :9153
        forward . /etc/resolv.conf {
          prefer_udp
        }
        k8s_external example.org
        cache 30
        loop
        reload
        loadbalance
    }
```

Following is coredns exposed as lb service:

```
[root@node1 ~]#
[root@node1 ~]# kubectl get svc -n kube-system
NAME      TYPE           CLUSTER-IP   EXTERNAL-IP    PORT(S)                                    AGE
coredns   LoadBalancer   10.233.0.3   10.41.12.138   53:31757/UDP,53:31720/TCP,9153:31461/TCP   28h
[root@node1 ~]#
[root@node1 ~]#
```

Following is a sample lb service that I have:

```
[root@node1 ~]# kubectl get svc
NAME         TYPE           CLUSTER-IP                  EXTERNAL-IP                     PORT(S)        AGE
kubernetes   ClusterIP      10.233.0.1                  <none>                          443/TCP        29h
my-nginx     LoadBalancer   fd85:ee78:d8a6:8607::14f3   2001:5b0:ffff:12:10:41:12:138   80:30561/TCP   8h
```

Now when I try to query the above service using external zone from a host outside the cluster it works as expected:
```
[root@RH8TESTK8VM01 ~]# dig @10.41.12.138 my-nginx.default.example.org AAAA +short
2001:5b0:ffff:12:10:41:12:138
```
But my expectation is that if I try to query the same service using its in cluster fqdn, coredns should not return anything instead it returns the in cluster ip for the service:

```
[root@RH8TESTK8VM01 ~]# dig @10.41.12.138 my-nginx.default.svc.cluster.local AAAA +short
fd85:ee78:d8a6:8607::14f3
```

I dont see this issue with the k8s_gateway plugin; is there some additional config required to stop coredns to respond to outside queries for in cluster services with in cluster ip's ?

**Comment by chrisohaver** [comment]
That's how it works.  K8s_external requires an active kubernetes plugin to function - it uses the kubernetes plugin's connection to the k8s API.   Since the kubernetes plugin is active, it will answer queries for the zone it is configured to serve.  

k8s_gateway works differently because it makes its own connection to the k8s API.


If you _only_ want to serve the example.org zone, make the server block only serve example.org (first line of Corefile)...

```
  Corefile: |
    example.org:53 {
        errors
        health {
            lameduck 5s
        }
        ready
        kubernetes cluster.local in-addr.arpa ip6.arpa {
          pods insecure
          fallthrough in-addr.arpa ip6.arpa
        }
        prometheus :9153
        forward . /etc/resolv.conf {
          prefer_udp
        }
        k8s_external example.org
        cache 30
        loop
        reload
        loadbalance
    }
```

If you want to still serve cluster.local to internal clients, you may be able to use the _acl_ plugin ... e.g. to only allow `cluster.local.` zone queries for clients in your Pod subnet. Something like... 

```
  Corefile: |
    .:53 {

        acl cluster.local. {
            allow net <pod-cidrs>
            block
        }

        errors
        health {
            lameduck 5s
        }
        ready
        kubernetes cluster.local in-addr.arpa ip6.arpa {
          pods insecure
          fallthrough in-addr.arpa ip6.arpa
        }
        prometheus :9153
        forward . /etc/resolv.conf {
          prefer_udp
        }
        k8s_external example.org
        cache 30
        loop
        reload
        loadbalance
    }
```

**Reply by micksey9** [reply]
hey chris, thanks a lot for explaining. I tried configuring an acl in the coredns config but it doesn't seem to work, coredns is still responding with in cluster ip's for out of cluster queries. This is what I configured in my corefile:

-----
  Corefile: |

    .:53 {

        acl cluster.local {
            allow net 10.233.64.0/18 fd85:ee78:d8a6:8607::1:0/112
            block
        }
        
        log
        errors
        health {
            lameduck 5s
        }
        ready
        kubernetes cluster.local in-addr.arpa ip6.arpa {
          pods insecure
          fallthrough in-addr.arpa ip6.arpa
        }
        prometheus :9153
        forward . /etc/resolv.conf {
          prefer_udp
        }
        k8s_external example.org
        cache 30
        loop
        reload
        loadbalance
    }

---- 
Even tried allowing just IPv4 POD CIDR and using block attribute to block queries from a subnet but it's not working.
[root@RH8TESTK8VM01 ~]# dig @10.41.12.138 my-nginx.default.svc.cluster.local A +short
10.233.62.33
[root@RH8TESTK8VM01 ~]# dig @10.41.12.138 my-nginx.default.svc.cluster.local AAAA +short
fd85:ee78:d8a6:8607::1256

----
Another weird thing I'm noticing is that every alternate query is going unanswered after i configured the acl:

[root@RH8TESTK8VM01 ~]# dig @10.41.12.138 my-nginx.default.svc.k8cluster.local A +short
[root@RH8TESTK8VM01 ~]# dig @10.41.12.138 my-nginx.default.svc.k8cluster.local A +short
10.233.62.33
[root@RH8TESTK8VM01 ~]# dig @10.41.12.138 my-nginx.default.svc.k8cluster.local AAAA +short
[root@RH8TESTK8VM01 ~]# dig @10.41.12.138 my-nginx.default.svc.k8cluster.local AAAA +short
fd85:ee78:d8a6:8607::1256

**Comment by chrisohaver** [comment]
I would have expected the acl to work.  But I haven't tested it in this configuration.  What IP does the coredns log say the queries come from?

If you don't need pods to be able to resolve example.org, then another way to solve it would be to split it out into a separate deployment. That is, leave the the cluster DNS as default, only used to resolve in-cluster pods.  And run a new deployment with k8s_external + kubertnetes.

Ultimately, if trying to do this in one deployment, it needs to implement "split dns"(where different classes of clients get different answers, e.g. depending on what ). Also called "split horizon dns".  This is hard to do in single CoreDNS deployment, and generally it's more flexible to split things up into separate CoreDNS's.

**Comment by johnbelamaric** [comment]
External LB is probably talking to the nodeport, and the IP seen from
CoreDNS is NAT'd:

https://kubernetes.io/docs/tutorials/services/source-ip/

**Reply by micksey9** [reply]
hey @chrisohaver @johnbelamaric , thanks for responding; following are the logs and yes it seems like when queried using the external LB ip (10.41.12.138), coredns sees the NAT'd source ip:

[INFO] Reloading complete
[INFO] 127.0.0.1:46575 - 15329 "HINFO IN 7594203555243935541.7524057424582865341. udp 57 false 512" NXDOMAIN qr,rd,ra 132 0.003810834s
[INFO] 10.233.92.0:51867 - 39333 "A IN my-nginx.default.svc.k8cluster.local. udp 77 false 4096" NOERROR qr,aa,rd 106 0.000192147s
[INFO] 10.233.96.0:19915 - 30040 "A IN grafana.com.monitoring.svc.k8cluster.local. tcp 60 false 65535" NXDOMAIN qr,aa,rd 159 0.000166411s
[INFO] 10.233.96.0:19915 - 30488 "A IN grafana.com.svc.k8cluster.local. tcp 49 false 65535" NXDOMAIN qr,aa,rd 148 0.000116223s
[INFO] 10.233.96.0:19915 - 51165 "AAAA IN grafana.com.k8cluster.local. tcp 45 false 65535" NXDOMAIN qr,aa,rd 144 0.000068594s
[INFO] 10.233.96.0:19915 - 45278 "A IN raw.githubusercontent.com.monitoring.svc.k8cluster.local. tcp 74 false 65535" NXDOMAIN qr,aa,rd 173 0.000196546s
[INFO] 10.233.96.0:19915 - 55454 "AAAA IN raw.githubusercontent.com.svc.k8cluster.local. tcp 63 false 65535" NXDOMAIN qr,aa,rd 162 0.000130224s
[INFO] 10.233.96.0:19915 - 19663 "AAAA IN raw.githubusercontent.com.k8cluster.local. tcp 59 false 65535" NXDOMAIN qr,aa,rd 158 0.000106334s
[INFO] 10.233.92.0:32799 - 52447 "A IN my-nginx.default.svc.k8cluster.local. udp 77 false 4096" NOERROR qr,aa,rd 106 0.000096825s
[INFO] 10.233.92.0:58036 - 48441 "A IN my-nginx.default.svc.k8cluster.local. udp 77 false 4096" NOERROR qr,aa,rd 106 0.000133136s
[INFO] 10.41.12.131:12142 - 56928 "AAAA IN my-nginx.default.svc.k8cluster.local. udp 77 false 4096" NOERROR qr,aa,rd 118 0.000149448s
[INFO] 10.41.12.131:28338 - 44574 "AAAA IN my-nginx.default.svc.k8cluster.local. udp 77 false 4096" NOERROR qr,aa,rd 118 0.000069082s
[INFO] 10.41.12.131:26847 - 12685 "A IN my-nginx.default.svc.k8cluster.local. udp 77 false 4096" NOERROR qr,aa,rd 106 0.00018853s
[INFO] 10.41.12.131:41040 - 11330 "A IN my-nginx.default.svc.k8cluster.local. udp 77 false 4096" NOERROR qr,aa,rd 106 0.000094322s
[INFO] **10.233.92.0:25334** - 7885 "A IN my-nginx.default.svc.k8cluster.local. udp 77 false 4096" NOERROR qr,aa,rd 106 0.000234264s
[INFO] **10.233.92.0:58155** - 63685 "A IN my-nginx.default.svc.k8cluster.local. udp 77 false 4096" NOERROR qr,aa,rd 106 0.000279721s

Highlighted is what coredns is seeing when queried on the external lb ip from outside the cluster.

**Comment by chrisohaver** [comment, marked as answer]
> the IP seen from CoreDNS is NAT'd:

Ah - thanks.  I suspected, which is why I asked what IP the source was in coredns log.

Splitting the deployments up is probably the best bet then.  In that new deployment, if both k8s_external and kubernetes plugin use the same zones (`example.org`), only k8s_external will answer, since it is executed before kubernetes (in `/plugin.cfg`).

**Reply by micksey9** [reply]
Ok will test that out. Thanks a lot for your help.

**Comment by anselmobattisti** [comment]
Thanks for sharing!

In my case i was tring to access the service using the same FQN for the local.cluster

service.default.svc.example.org

however, the correct FQN is 

service.default.example.org
